ref(pkg/chartutil): remove k8s version object dependency

adamreese opened the pull request:
Flattens the `.Capabilities` built-in and removes useless kubernetes
runtime metadata.

bacongobbler commented on docs/faq.md:
```diff
@@ -15,6 +15,11 @@ Here's an exhaustive list of all the major changes introduced in Helm 3.
 
+### Capabilities
+
+Capabilities built-in has been simplified.
```
Would you mind documenting what fields changed, and to what? e.g. `.Capabilities.KubeVersion` to `.Capabilities.Version`. This is a major breaking change to existing charts so we should document the changes required to existing charts.

adamreese replied:
updated to be backwards compatible

bacongobbler replied:
Some of the field names did change (for the better) so we should probably still list out what fields changed, and where. For example:

- `.Capabilities.KubeVersion.GitVersion` -> `.Capabilities.KubeVersion.Version`

I don't believe the platform/runtime fields were flattened either. e.g.

- `.Capabilities.KubeVersion.Platform` -> `.Capabilities.Platform`

Are you planning to add those back?

hickeyma replied:
I agree with @bacongobbler that it would be good to show the changes. Easy for user to see then.

adamreese replied:
Platform was a garbage field that was sentenced to collection.

Some fields we never populated with relevant information.  I deleted all the no-op fields.

@bacongobbler `GitVersion` still returns the same value.  `.Capabilities.KubeVersion.GitVersion == .Capabilities.KubeVersion.Version`

mattfarina commented on pkg/chartutil/testdata/subpop/charts/subchart1/templates/service.yaml:
```diff
@@ -5,9 +5,9 @@ metadata:
-    kube-version/gitversion: "v{{ .Capabilities.KubeVersion.Major }}.{{ .Capabilities.KubeVersion.Minor }}.0"
+    kube-version/major: "{{ .Capabilities.Major }}"
+    kube-version/minor: "{{ .Capabilities.Minor }}"
+    kube-version/gitversion: "v{{ .Capabilities.Major }}.{{ .Capabilities.Minor }}.0"
```
If this isn't really the `gitversion` should we even include the label?

adamreese replied:
this is just testdata but I updated it for you @mattfarina

adamreese commented on docs/chart_template_guide/builtin_objects.md:
```diff
@@ -19,7 +19,9 @@ In the previous section, we use `{{.Release.Name}}` to insert the name of a rele
-  - `Capabilities.KubeVersion` provides a way to look up the Kubernetes version. It has the following values: `Major`, `Minor`, `GitVersion`, `GitCommit`, `GitTreeState`, `BuildDate`, `GoVersion`, `Compiler`, and `Platform`.
+  - `Capabilities.Version` is the Kubernetes version.
+  - `Capabilities.Major` is the Kubernetes major version.
+  - `Capabilities.Minor` is the Kubernetes minor version.
```
updated to be backwards compatible

bacongobbler commented on pkg/chartutil/capabilities.go:
```diff
@@ -16,89 +16,66 @@ limitations under the License.
+// GitVersion returns the Kubernetes version string.
+//
+// Depreciated: use KubeVersion.Version.
+func (kv *KubeVersion) GitVersion() string { return kv.Version }
```
oh! I see now. Awesome!

bacongobbler commented on pkg/chartutil/capabilities.go:
```diff
@@ -16,89 +16,66 @@ limitations under the License.
+
+// GitVersion returns the Kubernetes version string.
+//
+// Depreciated: use KubeVersion.Version.
```
small FYI: the correct term is "[deprecated](https://en.wikipedia.org/wiki/Deprecation)". :)

mattfarina commented on pkg/chartutil/capabilities.go:
```diff
@@ -16,89 +16,67 @@ limitations under the License.
+
+// GitVersion returns the Kubernetes version string.
+//
+// Deprecated: use KubeVersion.Version.
```
`KubeVersion.Version` isn't documented in the `builtin_objects.md` file. Can that be added alongside `Major` and `Minor`?

mattfarina commented on pkg/chartutil/capabilities.go:
```diff
@@ -16,89 +16,67 @@ limitations under the License.
+type KubeVersion struct {
+	Version string // Kubernetes version
+	Major   string // Kubernetes major version
+	Minor   string // Kubernetes minor version
```
This is where we get into the ugliness of versions and how they are misused.

I have a gke test cluster with a version of `1.12.7-gke.10`. It's the reason you can see things like `if semverCompare "^1.7-0" .Capabilities.KubeVersion.GitVersion` all over the charts repo. The `-0` on the end tells semver to look at pre-releases. gke is misusing the pre-release part of semver.

Anything after minor is no longer available. On one hand, that will simplify a what many chart authors have to do. On the other hand, do we know of anywhere the lost information will have an impact?

I just wanted to ask the question and document this.

mattfarina reviewed (changes requested):
I tried to use this branch to install a configmap into a cluster to see the values for `GitVersion` and `Version`. The cluster was 1.12. The values for both were 1.14.0. It looks like it's using the default values rather than the actual values for the cluster. We need the cluster values for this to be useful.

mattfarina reviewed (approved):
@adamreese This looks good but needs a rebase. My only outstanding issue with this was fixed via the other PR that caused the need for the rebase.

adamreese committed:
ref(pkg/chartutil): remove k8s version object dependency

Flattens the `.Capabilities` built-in and removes useless kubernetes
runtime metadata.

Signed-off-by: Adam Reese <adam@reese.io>
